### PR TITLE
add turbopack-image

### PR DESCRIPTION
**Pull request by sokra**
### Description

allows to have custom plugins for module types

add turbopack-image crate which adds some image processing and blur placholder generation

next.js PR: https://github.com/vercel/next.js/pull/48531

### CODE REVIEW

File: crates/turbopack-image/src/process.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+const BLUR_QUALITY: u8 = 70;
+
+#[turbo_tasks::function]
+pub async fn get_meta_data_and_blur_placeholder(content: FileContentVc) -> Result<ImageMetaDataVc> {
+    let FileContent::Content(content) = &*content.await? else {
+      bail!("Input image not found");
+    };
+    let bytes = content.content().to_bytes()?;
+    let reader = image::io::Reader::new(Cursor::new(&bytes));
+    let reader = reader.with_guessed_format()?;
```
Comment (sokra) [inline crates/turbopack-image/src/process.rs:37]: This leads to crashes when format cannot be guessed

File: crates/turbopack-image/src/process.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    let format = reader.format();
+    let image = reader.decode()?;
```
Comment (sokra) [inline crates/turbopack-image/src/process.rs:39]: Error should lead to emitted Issues and fallback data

File: crates/turbopack-image/src/process.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+use turbo_tasks_fs::{FileContent, FileContentVc, FileSystemPathVc};
+use turbopack_core::{
+    error::PrettyPrintError,
+    ident::AssetIdentVc,
+    issue::{Issue, IssueVc},
+};
+
+#[turbo_tasks::value]
+#[serde(rename_all = "camelCase")]
+pub struct ImageMetaData {
```
Comment (alexkirsz) [inline crates/turbopack-image/src/process.rs:24]: Comment the public items of this module.

File: crates/turbopack-image/src/process.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        Some(ImageFormat::Png)
+            | Some(ImageFormat::Jpeg)
+            | Some(ImageFormat::WebP)
+            | Some(ImageFormat::Avif)
+    ) {
+        fn compute_blur_data(
+            image: image::DynamicImage,
+            format: ImageFormat,
+        ) -> Result<(String, u32, u32)> {
+            let small_image = image.resize(BLUR_IMG_SIZE, BLUR_IMG_SIZE, FilterType::Triangle);
```
[the remark this answers is not in the document]
Comment (sokra) [reply crates/turbopack-image/src/process.rs:87]: I don't think it's worth it for a 8x8 image...

File: crates/turbopack-image/src/process.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+#[turbo_tasks::value]
+#[serde(rename_all = "camelCase")]
+pub struct ImageMetaData {
+    width: u32,
+    height: u32,
+    #[serde(rename = "blurDataURL")]
+    blur_data_url: Option<String>,
+    blur_width: u32,
+    blur_height: u32,
```
[the remark this answers is not in the document]
Comment (sokra) [reply crates/turbopack-image/src/process.rs]: moved it

**Review by alexkirsz — APPROVED** [review]

**Pull request merged**